### PR TITLE
Improve delete unused tags to be able to process in chunks

### PR DESCRIPTION
**Pull request by zdong2**
Similar to PR #1184, break down the tag deletion cleanup task to perform lookup and cleanup in smaller chunks.

### CODE REVIEW

File: genie-web/src/main/java/com/netflix/genie/web/data/services/PersistenceService.java [hunk 1]
```diff
@@ -1121,22 +1121,24 @@ <R extends CommonResource> void removeTagForResource(
...
+                          @NotNull Instant createdThresholdUpperBound,
+                          @Min(1) int batchSize);
     //endregion
 
     //region File APIs
 
     /**
-     * Delete all files from the database that aren't referenced which were created before the supplied created
-     * threshold.
+     * Delete all files from the database that aren't referenced which were created between the supplied thresholds.
```
Comment (yingyiz-netflix) [inline genie-web/src/main/java/com/netflix/genie/web/data/services/PersistenceService.java:1141]: minor: `thresholds` -> `threshold bounds`

File: genie-web/src/main/java/com/netflix/genie/web/data/services/impl/jpa/repositories/JpaTagRepository.java [hunk 2]
```diff
@@ -45,7 +45,8 @@ public interface JpaTagRepository extends JpaIdRepository<TagEntity> {
             + "AND id NOT IN (SELECT DISTINCT(tag_id) FROM commands_tags) "
             + "AND id NOT IN (SELECT DISTINCT(tag_id) FROM criteria_tags) "
             + "AND id NOT IN (SELECT DISTINCT(tag_id) FROM jobs_tags) "
-            + "AND created <= :createdThreshold "
+            + "AND created <= :createdThresholdUpperBound "
+            + "AND created >= :createdThresholdLowerBound "
```
[the remark this answers is not in the document]
Comment (yingyiz-netflix) [reply genie-web/src/main/java/com/netflix/genie/web/data/services/impl/jpa/repositories/JpaTagRepository.java:49]: nvm, I see your comment

File: genie-web/src/main/java/com/netflix/genie/web/tasks/leader/DatabaseCleanupTask.java [hunk 3]
```diff
@@ -411,19 +414,36 @@ private void deleteTags(final Instant creationThreshold) {
...
+                Instant upperBound = creationThreshold;
+                Instant lowerBound = creationThreshold.minus(rollingWindowHours, ChronoUnit.HOURS);
+                final Instant batchLowerBound = creationThreshold.minus(batchDaysWithin, ChronoUnit.DAYS);
+                while (upperBound.isAfter(batchLowerBound)) {
+                    totalDeleted += deleteUnusedTagsBetween(lowerBound, upperBound, batchSize);
+                    upperBound = lowerBound;
+                    lowerBound = lowerBound.minus(rollingWindowHours, ChronoUnit.HOURS);
+                }
+                // do a final deletion of everything <= batchLowerBound
+                totalDeleted += deleteUnusedTagsBetween(Instant.EPOCH, upperBound, batchSize);
```
Comment (yingyiz-netflix) [inline genie-web/src/main/java/com/netflix/genie/web/tasks/leader/DatabaseCleanupTask.java:446]: from `epoch` to `upperBound` can be a large range, I think it's almost [epoch, today - 30 days), would this be a concern?

File: genie-web/src/main/java/com/netflix/genie/web/data/services/impl/jpa/repositories/JpaTagRepository.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yingyiz-netflix) [reply genie-web/src/main/java/com/netflix/genie/web/data/services/impl/jpa/repositories/JpaTagRepository.java:49]: wondering if we can move the `created <= :createdThresholdUpperBound` and  `created >= :createdThresholdLowerBound "` to the beginning of WHERE clause.

**Review by yingyiz-netflix — APPROVED** [review]

File: genie-web/src/main/java/com/netflix/genie/web/data/services/impl/jpa/repositories/JpaTagRepository.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (zdong2) [reply genie-web/src/main/java/com/netflix/genie/web/data/services/impl/jpa/repositories/JpaTagRepository.java:49]: Maybe we can keep it, as I used `EXPLAIN` to compare the query execution plans and found them to be identical.

File: genie-web/src/main/java/com/netflix/genie/web/tasks/leader/DatabaseCleanupTask.java (hunk 3 above)
Comment (zdong2) [reply genie-web/src/main/java/com/netflix/genie/web/tasks/leader/DatabaseCleanupTask.java:446]: It shouldn't be a concern, as the `batchSize` still limits the number of rows it processes and deletes in each iteration.

**Review by yingyiz-netflix — APPROVED** [review]

**Pull request merged**